---
날짜: 2024-05-12
넘버: 
태그: 프로그래밍
출처: "[Udemy]완전 초보자를 위한 Java 프로그래밍 : 단기간에 Java 완벽 정복"
aliases:
---
### 날짜:  2024-05-12 15:20

### 태그: #프로그래밍 

>[!메모] 예외 처리 팁
> - 예외가 발생하면 stack trace를 전부 log에 넣자.
> - flow control을 위해 사용하지 마라. (if/else보다 비용이 높다)
> - 사용자를 염두에 두어라 (사용자는 무엇을 보고 싶어하겠는가)
> - 에러를 직접 다룰 support 팀을 고려하라.
> - 호출 메서드를 고려하라. Checked Exception으로 넘겼는데, 이미 호출한 메서드가 할 수 있는 것이 없다면, 잘못되어 있는 것이다. (Spring의 좋은점: 대부분의 예외를 RuntimeException으로만 다루고 있음)
> 	- 호출 메서드가 예외에 대해 반응하도록 하는 방법이 없는가? -> 없다면 Runtime으로..
> - global exception handling을 하라. 사용자는 예외를 직접 보아서는 안 된다. 예외에 대한 출력만 볼 수 있도록 하라
> 


### 원문
---
# 예외 처리
- 예외를 처리하는 2가지 키
	- 사용자에게 메시지를 남기는 것.
	- 어떤 작업을 거쳐야하는지 안내하는 것.
### Stack Trace
- 어떤 스택에서 발생한 에러인지 알려준다.
- 해당 스택에서 처리되지 않은 예외는 calling method로 넘겨진다.
```java title:"stack trace" error:27-31
package ExceptionHandling;

public class ExceptionHandlingRunner {  
	public static void main(String[] args) {  
		method1();
		System.out.println("main");
	}  

	private static void method1() {  
		method2();
		System.out.println("method1");
	}

	private static void method2() {  
		method3();
		System.out.println("method2");
	}

	private static void method3() {  
		String str = null;  
		str.length();  
		System.out.println("method3");
	}  
}

/* stack trace
Exception in thread "main" java.lang.NullPointerException: Cannot invoke "String.length()" because "str" is null
	at ...method3(ExceptionHandlingRunner.java:18)
	at ...method2(ExceptionHandlingRunner.java:13)
	at ...method1(ExceptionHandlingRunner.java:9)
	at ...main(ExceptionHandlingRunner.java:5)
*/
```
- 예외가 처리되지 않으면 다음 코드로 진행하지 않는다.
## try/catch 블록
### Exception 
- Exception 클래스의 주석
> The class Exception and its subclasses are a form of **Throwable** that indicates conditions that a **reasonable application might want to catch.**
- Throwable 클래스를 상속받았다.
- 합리적인 애플리케이션이 catch 할 수 있는 상태를 나타낸다.
```java hl:17
package ExceptionHandling;  
  
public class ExceptionHandlingRunner2 {  
	public static void main(String[] args) {  
		method1();  
	}  
	  
	private static void method1() {  
		method2();  
	}  
		  
	private static void method2() {  
		  
		try {  
			String str = null;  
			str.length();  
		} catch (Exception ex) {  
			System.out.println("Error caught");
		}  
	}  
}
```
### 예외별 catch 블럭
- 매칭되는 예외의 catch블럭을 실행한다.
	- 매칭되는 예외가 없다면 catch 블럭을 실행하지 않는다.
```java
try {  
	String str = null;  
	str.length();  
} catch (NullPointerException ex) {  
	System.out.println("Null pointer Error caught");
} catch (Exception ex) {  
	System.out.println("Error caught");
}  
```
### finally 블럭
- 리소스를 정리해줘야 하는 필요가 있을 때, finally 블록을 사용하여 try/catch 블록 이후에 무조건 실행하도록 할 수 있다.
```java
public static void main(String[] args) {  
	Scanner scanner = null;  
	try {  
		scanner = new Scanner(System.in);  
		int[] numbers = {12, 3, 4, 5};  
		int number = numbers[5];  
	} catch (Exception ex) {  
		ex.printStackTrace();  
	} finally {  
		System.out.println("Scanner closed");  
		if (scanner != null)
			scanner.close();  
	}  
}
```
### try-with-resources
- 자바 7부터 사용가능
- `AutoClosable <- Closable` 인터페이스를 구현한 클래스만 사용이 가능하다.
- resource를 자동으로 닫아준다.
- catch와 finally를 안 써줘도 된다.
```java
	try (Scanner scanner = new Scanner(System.in)) {  
		int[] numbers = {12, 3, 4, 5};  
		int number = numbers[5];  
	}
```
---
### 생각(파생된 질문/생각)

### 출처
- [완전 초보자를 위한 Java 프로그래밍 : 단기간에 Java 완벽 정복](https://www.udemy.com/course/best-java-programming/?couponCode=ST6MT42324)

### 연결 문서 (연결 이유)
